aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -2987,98 +3018,165 @@ std::tuple<Tensor,Tensor> eig(const Tensor& self, bool eigenvectors) {
 
-  AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(self.scalar_type(), "svd_cpu", [&]{
-    apply_svd<scalar_t>(self_working_copy, U_working_copy, S_working_copy, VT_working_copy, jobz, infos);
-  });
+  if (!U_ready) {
+    U.copy_(*U_);
+  }
+  if (!S.is_same(*S_)) {
+    S.copy_(*S_);
+  }
+  if (!Vh_ready) {
+    Vh.copy_(*Vh_);
+  }
 
-  if (self.dim() > 2) {
-    batchCheckErrors(infos, "svd_cpu");
+  // TODO This should be removed, and the code checking for convergence should be lifted
+  // from svd_cusolver to this function. We should then make sure that this function
+  // never errors out.
+  if (A.dim() > 2) {
+    batchCheckErrors(info, "linalg.svd");
   } else {
-    singleCheckErrors(infos[0], "svd_cpu");
+    singleCheckErrors(info.item<int64_t>(), "linalg.svd");
   }
+}
+
+std::tuple<Tensor&, Tensor&, Tensor&>
+linalg_svd_out(const Tensor& A,
+               bool full_matrices,
+               Tensor & U,
+               Tensor & S,
+               Tensor & Vh) {
+  // This function does not have an _ex variant as we always check errors inside
+  // to assure the convergence of the algorithm anyway. See
+  // https://github.com/pytorch/pytorch/issues/28293
+  // https://github.com/pytorch/pytorch/issues/64237
+  //
+  // We must delegate both linalg_svd and linalg_svdvals to
+  // _linalg_svd (rather than delegating linalg_svdvals to linalg_svd) because
+  //   1. We don't want to expose the `compute_uv` parameter in svd
+  //   2. We would like to make use of the `compute_uv=False` optimisation within svdvals
+  // The only way to achieve these two things and still abide by the compositionality rules
+  // is by dispatching to another function.
+  return at::_linalg_svd_out(U, S, Vh, A, full_matrices, /*compute_uv=*/true);
+}
+
+std::tuple<Tensor, Tensor, Tensor> linalg_svd(const Tensor& A, bool full_matrices) {
+  return at::_linalg_svd(A, full_matrices, /*compute_uv=*/true);
+}
+
+// See note in linalg_svd for why this function does not have an _ex variant
+Tensor& linalg_svdvals_out(const Tensor& A, Tensor & S) {
+  // Dummies
+  auto U = at::empty({0}, A.options());
+  auto Vh = at::empty({0}, A.options());
+  at::_linalg_svd_out(U, S, Vh, A, /*full_matrices=*/false, /*comptue_uv=*/false);
+  return S;
+}
+
+Tensor linalg_svdvals(const Tensor& A) {
+  const bool A_requires_grad = (at::GradMode::is_enabled() && A.requires_grad());
+  return std::get<1>(at::_linalg_svd(A, /*full_matrices=*/false, /*comptue_uv=*/A_requires_grad));
+}
+
+std::tuple<Tensor&, Tensor&, Tensor&> svd_out(const Tensor& self, bool some, bool compute_uv, Tensor& U, Tensor& S, Tensor& V) {
 
   if (compute_uv) {
-    // so far we have computed VT, but torch.svd returns V instead. Adjust accordingly.
-    // Note that the 'apply_svd' routine returns VT = V^T (for real inputs) or VT = V^H (for complex inputs), not V.
-    VT_working_copy = VT_working_copy.conj();
-    VT_working_copy.transpose_(-2, -1);
+    if (V.dim() >= 2) {
+      V.transpose_(-2, -1);
+    }
```

Comment:
What happens if V is a column-major matrix?

Reply:
This simply handles the size of `V`. If the strides are incorrect, `linalg_svd_out` (in relaity, `_linalg_svd_out`) will handle them.